app/helpers/alchemy/admin/elements_helper.rb

```diff
@@ -54,30 +54,42 @@ def elements_for_select(elements)
         end
       end
 
-      # Returns all elements that could be placed on that page because of the pages layout.
+      # Returns all elements that can be placed on the current page.
       # The elements will be grouped by cell.
+      #
+      # @param [Array] elements
+      #   collection of element objects
+      # @param [String] object_method
+      #   method that is called on the element objects used for the select option value
+      #
       def grouped_elements_for_select(elements, object_method = 'name')
         return "" if elements.blank?
         cells_definition = @page.cell_definitions
         return "" if cells_definition.blank?
         options = {}
-        celled_elements = []
         cells_definition.each do |cell|
-          cell_elements = elements.select { |e| cell['elements'].include?(e.class.name == 'Element' ? e.name : e['name']) }
-          celled_elements += cell_elements
+          cell_elements = elements_for_cell(elements, cell)
           optgroup_label = Cell.translated_label_for(cell['name'])
           options[optgroup_label] = cell_elements.map do |e|
             element_array_for_options(e, object_method, cell)
           end
         end
-        other_elements = elements - celled_elements
-        unless other_elements.blank?
-          options[_t(:main_content)] = other_elements.map do |e|
-            element_array_for_options(e, object_method)
-          end
+        options[_t(:main_content)] = elements_for_main_content(elements).map do |e|
+          element_array_for_options(e, object_method)
+        end
+        options.delete_if { |cell, elements| elements.blank? } # Throw out empty cells
```

Comment:
Shadowing outer local variable - `elements`.<br>Unused block argument - `cell`. If it's necessary, use `_` or `_cell` as an argument name to indicate that it won't be used.